### PR TITLE
Turned the backend into a QML plugin.

### PR DESCRIPTION
**Pull request by Druage**
This now turns the project into a QML plugin. Statically linking to this backend has been removed, but canquickly added by anyone. ( It's not hard. )

### CODE REVIEW

**Comment by Druage** [comment]
We cannot merge this PR, until the "feature-importer-mt" branch of Phoenix, has been merged into its "master" branch because this "qmlplugin" branch was based off of the "feature-importer-mt" branch.

**Comment by Druage** [comment]
I forgot about how the frontend is synced to a specific commit in master, so this won't affect that.

File: backend.pro [hunk 1]
```diff
@@ -1,78 +1,91 @@
...
 
-    # FIXME: Remove once newer Qt versions make this unnecessary
-    macx: QMAKE_MAC_SDK = macosx10.11
+____________________________________________________________________
+#################
+# Include Files #
+#################
 
-    # Include libraries
+    # Include platform specific libraries
```
Comment (athairus) [inline backend.pro:40]: I think it looked neater before...

File: backend.pro [hunk 2]
```diff
@@ -1,78 +1,91 @@
-##
-## Extra targets
-##
+#################
+# Extra targets #
```
Comment (athairus) [inline backend.pro]: You And I Are Different When It Comes To Section Headers, I don't like to use Title Case (and for some reason you didn't use it here lol)

File: backendplugin.cpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
+#include "backendplugin.h"
```
Comment (athairus) [inline backendplugin.cpp]: Don't include anything in .cpp files other than their corresponding header!

File: backendcommon.h [hunk 4]
```diff
@@ -14,10 +14,6 @@
 
 #include <math.h>
 
-#include "samplerate.h"
```
Comment (athairus) [inline backendcommon.h:17]: Why were these removed? Are you not using the precompiled header anymore?

File: backendplugin.h [hunk 5]
```diff
@@ -0,0 +1,14 @@
+#ifndef COREPLUGIN_H
+#define COREPLUGIN_H
+
+#include <QQmlExtensionPlugin>
```
Comment (athairus) [inline backendplugin.h]: Use the common header

File: consumer/audiooutput.cpp [hunk 6]
```diff
@@ -1,5 +1,5 @@
-
 #include "audiooutput.h"
+#include "logging.h"
```
Comment (athairus) [inline consumer/audiooutput.cpp]: Don't include anything in the .cpp file besides the class's header

File: consumer/audiooutput.h [hunk 7]
```diff
@@ -5,11 +5,9 @@
 
 #include "consumer.h"
 #include "controllable.h"
-
+#include <samplerate.h>
```
Comment (athairus) [inline consumer/audiooutput.h]: Use precompiled header, separate from our own headers (put `#include "backendcommon.h"` above `#include "consumer.h"` with a blank line)

File: consumer/videooutput.cpp [hunk 8]
```diff
@@ -32,6 +32,9 @@ VideoOutput::~VideoOutput() {
 }
 
 void VideoOutput::consumerFormat( ProducerFormat format ) {
+
+    qDebug() << "CONSUMER FORMAT START";
```
Comment (athairus) [inline consumer/videooutput.cpp]: Didn't you make a big stink a long time ago about making sure to tag debug messages? :)

File: core/core.cpp [hunk 9]
```diff
@@ -1,5 +1,8 @@
 #include "core.h"
 
+#include "logging.h"
```
Comment (athairus) [inline core/core.cpp:3]: Don't include anything in the .cpp file besides the class's header

File: core/core.h [hunk 10]
```diff
@@ -3,8 +3,6 @@
 
 #include "backendcommon.h"
 
-#include "logging.h"
```
Comment (athairus) [inline core/core.h]: Good, this should go in backendcommon anyway

File: core/corecontrolproxy.cpp [hunk 11]
```diff
@@ -1,5 +1,8 @@
 #include "corecontrolproxy.h"
 
+#include <type_traits>
+#include <QQmlApplicationEngine>
```
Comment (athairus) [inline core/corecontrolproxy.cpp:8]: Don't include anything in the .cpp file besides the class's header

File: core/corecontrolproxy.h [hunk 12]
```diff
@@ -59,10 +59,10 @@ class CoreControlProxy : public QObject {
         // Producers and consumers that live in QML (and the QML thread)
 
         // Needed by: "libretro"
-        Q_PROPERTY( VideoOutput *videoOutput READ getVideoOutput WRITE setVideoOutput NOTIFY videoOutputChanged )
+        Q_PROPERTY( QObject *videoOutput READ getVideoOutput WRITE setVideoOutput NOTIFY videoOutputChanged )
```
Comment (athairus) [inline core/corecontrolproxy.h]: Why are these changed?

File: input/inputmanager.cpp [hunk 13]
```diff
@@ -1,18 +1,23 @@
 #include "inputmanager.h"
 
+#include "joystick.h"
```
Comment (athairus) [inline input/inputmanager.cpp]: Don't include anything in the .cpp file besides the class's header

File: input/inputmanager.h [hunk 14]
```diff
@@ -6,13 +6,13 @@
...
 #include "producer.h"
 
-#include "qmlinputdevice.h"
-#include "sdleventloop.h"
-#include "inputdevice.h"
-#include "keyboard.h"
-
 #include "logging.h"
 
+class InputDevice;
```
Comment (athairus) [inline input/inputmanager.h:7]: I can see you doing your forward declarations here. For consistency's sake, we should use my style.

File: input/sdleventloop.cpp [hunk 15]
```diff
@@ -1,10 +1,13 @@
 #include "sdleventloop.h"
-
 #include "logging.h"
+#include <SDL.h>
```
Comment (athairus) [inline input/sdleventloop.cpp:6]: Don't include anything in the .cpp file besides the class's header

File: input/sdleventloop.h [hunk 16]
```diff
@@ -1,11 +1,13 @@
 #ifndef SDLEVENTLOOP_H
 #define SDLEVENTLOOP_H
 
-#include "backendcommon.h"
-
-#include "joystick.h"
+class Joystick;
 
 // The SDLEventLoop's job is to poll for button states and to emit signals as devices are connected or disconnected.
+#include <QObject>
```
Comment (athairus) [inline input/sdleventloop.h:4]: Use the precompiled header

File: util/logging.h [hunk 17]
```diff
@@ -2,6 +2,7 @@
 #define LOGGING_H
 
 #include "backendcommon.h"
+#include <QDebug>
```
Comment (athairus) [inline util/logging.h:4]: Should be covered by precompiled header

File: backend.pro [hunk 18]
```diff
@@ -96,39 +109,42 @@
         role/control.cpp \
         role/producer.cpp \
         util/logging.cpp \
-        util/looper.cpp \
-    cmdlineargs.cpp
+        util/looper.cpp  \
+        backendplugin.cpp
+____________________________________________________________________
+####################
+# Linkers Settings #
```
Comment (athairus) [inline backend.pro]: Linker

File: consumer/videooutput.cpp (hunk 8 above)
Comment (Druage) [inline consumer/videooutput.cpp]: Yes, this was meant to be deleted.
